### PR TITLE
Readds map diversity

### PR DESCRIPTION
**Pull request by kriskog**
## About The Pull Request
Changes map vote to only allow for changing the next map once per round, and adds map tracking of the last two rounds to SSPersistence. Map votes will no longer include maps that have been chosen twice in the last three rounds, including current. 

In practice, this means maps like Metastation which have a ridiculous play rate (>95%) should land at a maximum of 50% play rate if votes control the map choice all the time. Map rotation choice is not affected by this, but will be unable to change away from voted map.

To satisfy my curiosity I grabbed some map distribution data from our logs.
https://cdn.discordapp.com/attachments/326831214667235328/661007980032163845/unknown.png

## Why It's Good For The Game
We have way too little played map diversity, and it is hurting both our replayability and the state of our map pool. This should hopefully also get people working on our other maps again so we don't end up with a ton of unused outdated maps.

## Changelog
:cl: Skoglol
tweak: Map vote now only allows one map vote/rotation per round. Maps that have been played twice in the last three rounds including current are not available for voting.
/:cl:

### CODE REVIEW

**Comment by optimumtact** [comment]
Approved for merge.

part of actually getting good maps means playing the whole rotation, not just the one most popular

**Comment by Borisvanmemes** [comment]
Thank you for this.

File: code/controllers/subsystem/persistence.dm [hunk 1]
```diff
@@ -104,6 +106,15 @@ SUBSYSTEM_DEF(persistence)
 		return
 	saved_modes = json["data"]
 
+/datum/controller/subsystem/persistence/proc/LoadRecentMaps()
+	var/json_file = file("data/RecentMaps.json")
```
Comment (Cyberboss) [inline code/controllers/subsystem/persistence.dm]: Define for this, also do we not have a variable for `data/`?

**Comment by wesoda25** [comment]
finally

File: code/controllers/subsystem/persistence.dm [hunk 2]
```diff
@@ -104,6 +106,15 @@ SUBSYSTEM_DEF(persistence)
...
+	if(!fexists(json_file))
```
Comment (Cyberboss) [inline code/controllers/subsystem/persistence.dm]: `fexists` takes a string.

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/persistence.dm (hunk 1 above)
Comment (Cyberboss) [reply code/controllers/subsystem/persistence.dm]: We don't

Comment (kriskog) [reply code/controllers/subsystem/persistence.dm]: Heck, copy pasted LoadRecentModes but didnt scrutinize it. Ill fix tomorrow

**Comment by TheVekter** [comment]
Oh god, _finally_. This is going to upset people, but fuck it, playing Meta every round is boring.

**Comment by trollbreeder** [comment]
Thank christ. I was getting sick of metastation because voterbus loves to dump all votes into either kilo gang or meta conglomerate.
The only few times i got something voted for other than meta is when i rallied everyone to vote for delta, or when a map vote appeared constantly chanting "KILO GANG!". If it takes that much effort, then something needs to be drastically changed.

Oh by the way make this a config option please.

**Comment by BadSS13Player** [comment]
All maps other than Metastation suck because it's not Metastation.

**Comment by trollbreeder** [comment]
> > If it takes that much effort, then something needs to be drastically changed.
> 
> All maps other than Metastation suck because it's not Metastation.

Metastation is the dust2 of TG

**Comment by Michael-Ashfield** [comment]
Could I also suggest removing the preferred map option, it currently makes actually voting for any map other than meta useless. I have literally never seen a map other than meta win the vote, even if another map has more real votes.

**Comment by kriskog** [comment]
Default map is unset on the servers, and have been since late summer. Having your preferred map set to default results in a null vote. Therefore, only people who have consciously set their preferred map and no-votes count towards map votes.

**Comment by kriskog** [comment]
image

File: code/controllers/subsystem/persistence.dm (hunk 1 above)
Comment (kriskog) [reply code/controllers/subsystem/persistence.dm]: Would have fixed the rest while I was at it, but since you're touching this in #48448 I decided not to.

**Comment by BadSS13Player** [comment]
inb4 "This is a code solution to a player problem"

File: code/controllers/subsystem/persistence.dm [hunk 3]
```diff
@@ -270,6 +283,15 @@ SUBSYSTEM_DEF(persistence)
 	fdel(json_file)
 	WRITE_FILE(json_file, json_encode(file_data))
 
+/datum/controller/subsystem/persistence/proc/CollectMaps()
+	saved_maps[2] = saved_maps[1]
+	saved_maps[1] = SSmapping.config.map_name
+	var/json_file = file("data/RecentMaps.json")
```
Comment (Cyberboss) [inline code/controllers/subsystem/persistence.dm]: Missed this one

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/persistence.dm (hunk 3 above)
Comment (kriskog) [reply code/controllers/subsystem/persistence.dm]: Woops thanks

**Comment by CRITAWAKETS** [comment]
The graph shows a major decline for some reason in every map's appearance except for meta despite even the existance of a new map into the game.

_its like people think that meta is the best map_

**Comment by kriskog** [comment]
And yet
image

Note that map vote was made available to players during summer, so the "May" column is pre map vote.

**Review by Cyberboss — APPROVED** [review]

**Comment by Cyberboss** [comment]
Communism will prevail

**Comment by cacogen** [comment]
thank fucking christ

**Comment by trollbreeder** [comment]
hooray for democracy

**Pull request merged**